Render pipeline changes

Spongman opened the pull request:
ok, this is essentially the same PR that I submitted before (#2547), but with the thunk code removed.

it is rebased on top of https://github.com/processing/p5.js/pull/2587 since that is required to support the extended 3d state that is pushed.

Spongman commented:
71132faa4f2022818b84cde3b04bec2727d14481

Spongman commented:
ok, i have removed everythnig i can.

kjhollen commented:
Thanks @Spongman! It will take me some time to dig into the details here and grok how all the pieces fit together—I hope I can review later in the week.

Spongman commented:
here's a quick summary of some of the changes. hopefully it'll help:

- light.js: removing `setUniform` calls, replacing them with updates to renderer state
- material.js:  similarly, the shaders to use are determined during render, here we just update the state to indicate which ones to use.
- p5.Geometry.js: `dirtyFlags` indicates which buffers need to be updated. `reset` is used to re-initialize the render buffers.
- p5.RendererGL.Immediate.js: removed duplicate rendering code, now re-uses a `immediateMode` geometry. unifies uniform & buffer names between immediate & retained.
- p5.RendererGL.Retained.js: buffers are updated via loops over buffer definitions (`_prepareBuffers`). buffers are only updated if dirty.  all uniforms & blending are set in render methods. shaders are chosen in `drawBuffers`
- p5.RendererGL.js: initialization of new state, push/pop of said state. shader changes to handle uniform/buffer unification. factored out `lighting` calculation glsl code (`sumLights()`) into separate file.
- p5.Shader.js: handle changes to uniform/buffer names due to unification.

- *.frag: changes due to unification of uniform/attribute names. factored `sumLights()` lighting code. simplified texture-only shader, lit immediate-mode shader.

kjhollen commented on src/webgl/shaders/light.vert:
```diff
@@ -1,78 +1,69 @@
+// directional lights
 uniform int uDirectionalLightCount;
-uniform int uPointLightCount;
+uniform vec3 uDirectionalLightDirection[8];
```
Just a note that when this merges, we'll need to update the shader uniform and attribute documentation here: https://github.com/processing/p5.js/wiki/WebGL-Module-Architecture

kjhollen commented on src/webgl/p5.RendererGL.Retained.js:
```diff
@@ -320,18 +193,144 @@ p5.RendererGL.prototype.drawBuffersScaled = function(
+  fill.setUniform('uDirectionalLightCount', directionalLightCount);
+  fill.setUniform(
+    'uDirectionalLightDirection',
+    this.directionalLightDirections
```
this feels a little redundant and is quite a long name. can we avoid changing this and just note with an inline comment in the shader?

kjhollen replied:
I think uLightDirection is more concise and direct (hehe pun intended). :) I personally value that over consistency in this particular instance, but this isn't one of the more pressing issues in this patch.

kjhollen commented on src/webgl/p5.RendererGL.Immediate.js:
```diff
@@ -115,211 +92,136 @@ p5.RendererGL.prototype.endShape = function(
+
+        case constants.LINES:
+          drawMode = null;
+          break;
```
Why is this a null draw mode? all of the others record the draw mode here.

Spongman replied:
switch switch statement it selecting the draw mode to use for the fill. if the `shapeMode` is `LINES` then there's *no* (null) fill !

Spongman commented on src/webgl/light.js:
```diff
@@ -150,11 +137,19 @@ p5.prototype.ambientLight = function(v1, v2, v3, a) {
+};
 
+p5.RendererGL.prototype.directionalLight = function(v1, v2, v3, x, y, z) {
   //@TODO: check parameters number
```
i'm not sure about that comment, it was added in 65a2afd539874f09ab037529747abc8a2986a799

kjhollen replied:
cool, I see. after this is merged I'll update these comments to make it more clear that parameter validation is missing.

Spongman replied:
or we could just merge #2471 ;-)

Spongman commented on src/webgl/material.js:
```diff
@@ -168,13 +176,15 @@ p5.prototype.shader = function(s) {
 
+p5.RendererGL.prototype.normalMaterial = function() {
+  this.drawMode = constants.FILL;
+  this._enableNormal = true;
```
sure

Spongman commented on src/webgl/p5.RendererGL.Immediate.js:
```diff
@@ -88,19 +71,13 @@ p5.RendererGL.prototype.vertex = function(x, y) {
+  var push = Array.prototype.push;
+  im.vertices.push(new p5.Vector(x, y, z));
+  push.apply(im.vertexColors, this._diffuseColor);
+  push.apply(im.vertexAmbients, this._ambientColor);
```
this is purely a performance optimization that i made after profiling this code. the 2nd & 3rd calls are equivalient to:
````
Array.prototype.push(im.vertexColors, this._diffuseColor);
Array.prototype.push(im.vertexAmbients, this._ambientColor);
````
which appends the elements of the color arrays onto the vertex buffer arrays.

the `push` temporary is just there to save accessing the push method repeatedly. this was more of an issue originally when this code handled specularity, normals, ambients, etc...

kjhollen replied:
I understand what the code does, but I don't understand why the method isn't invoked the same way for all 4 cases. It's a strange thing to encounter side-by-side; a note about the optimization and why it's only needed in half of the cases would be helpful.

meiamsome replied:
The four cases are not the same, `push.apply` concatenates all the elements of an array onto the first argument whereas `push` just adds one element to an array:
```javascript
var a = [];
Array.prototype.push.apply(a, [1,2,3]);
console.log(a); // [1, 2, 3]
var b = [];
b.push([1, 2, 3]);
console.log(b); // [[1, 2, 3]]
```

It's equivalent to the spread operator here:
```javascript
var c = [];
c.push(...[1, 2, 3]);
console.log(c);
```
But we aren't using newer language features, so we have to use `push.apply`

Spongman replied:
it's only invoked like that in those two cases because those are the only two cases where the input elements are in an array.

Spongman replied:
spread also allocates a new array. which would be very bad here.

kjhollen replied:
got it, thank you.

kjhollen replied:
there still doesn't seem to be any reason that this can't be:

```
push.apply(im.vertices, new p5.Vector(x, y, z));
push.apply(im.vertexColors, this._diffuseColor);
push.apply(im.vertexAmbients, this._ambientColor);
push.apply(im.uvs, [u, v]);
```
unless the creation of the array on the last line is a performance hit? but could that just be noted with a comment? otherwise, interleaving the style choices still looks strange to me.

Spongman replied:
1) `push.apply(im.vertices, new p5.Vector(x, y, z))` is an invalid invocation of the `apply` method. the second argument needs to be an array
2) this is correct, but requires creating a new array - exactly what i'm trying to avoid here.

Spongman commented on src/webgl/p5.RendererGL.Immediate.js:
```diff
@@ -115,211 +92,136 @@ p5.RendererGL.prototype.endShape = function(
+          break;
+      }
+
+      im._edgesToVertices();
```
yeah, i had one, but it looks like it got lost in the rebase

Spongman replied:
i updated the immediateMode/basic test to reflect these.

Spongman commented on src/webgl/p5.RendererGL.Immediate.js:
```diff
@@ -115,211 +92,136 @@ p5.RendererGL.prototype.endShape = function(
+        if (!fill /*|| !fill.isColorShader()*/) {
+          fill = this._getImmediateFlatShader();
+        }
+      }
```
because in immediate mode there's no such thing as a 'texture shader' - it's just based on the `isTexture` uniform. this behavior hasn't changed - it's how it works now.

Spongman replied:
i never intended to unify the shaders completely (although the lighting calculation did get unified), simply because the requirements of the immediate and retained mode shaders are quite different. the main unification here is with the handling of the geometry buffers.

kjhollen replied:
okay. this is where having an issue filed that clearly states a feature or enhancement before a PR is submitted is helpful. please do this in the future!

Spongman replied:
well, to be fair, i _did_ list several pre-existing issues that the original PR addressed.

kjhollen replied:
yes, but a refactor of this scale was not necessary to fix the individual issues. an issue to document enhancing the architecture would have been helpful.

Spongman commented on src/webgl/p5.RendererGL.Retained.js:
```diff
@@ -194,100 +128,39 @@ p5.RendererGL.prototype.createBuffers = function(gId, obj) {
-      );
+  } else {
+    if (!fill /* || !fill.isColorShader()*/) {
+      fill = this._getColorShader();
```
not really. the main reason is that in immediate mode there's a richer set of per-vertex attributes. indeed, in the code that i removed before this PR the full set of per-vertex attributes from Processing is implemented. this requires a different set of shaders, and a different selection mechanism.

kjhollen replied:
okay, then let's have this as a separate method that can have tests written for immediate mode. thanks!

Spongman commented on src/webgl/p5.RendererGL.js:
```diff
@@ -100,28 +104,33 @@ p5.RendererGL = function(elt, pInst, isMainCanvas, attr) {
   this._defaultLightShader = undefined;
-  this._defaultImmediateModeShader = undefined;
+  this._defaultImmediateLightShader = undefined;
+  this._defaultImmediateFlatShader = undefined;
```
i didn't use 'color' because it could be textured - it's just not 'lit'.

Spongman replied:
also, the shaders for retained and immediate mode are different.

Spongman commented on src/webgl/shaders/immediateLight.vert:
```diff
@@ -0,0 +1,70 @@
+
+// material properties
+attribute vec3 aAmbientColor;
+attribute vec4 aMaterialColor;
```
nope, there's no way to do this dynamically. there's really not that much duplicated code - just the attribute/unfirom definitions. most of the lighting code is in `lighting.glsl`.

kjhollen replied:
I think it could be done if both parameters weren't called materialColor, but I have to think about this a little more when I can give it my full attention. I guess that could be done in a future patch but I'd like the public API for the shader uniforms and attributes to be as stable as possible.

Spongman commented on test/unit/webgl/p5.Shader.js:
```diff
@@ -176,70 +152,6 @@ suite('p5.Shader', function() {
-        "_renderer's curFillShader did not get set to light shader " +
-          'after call to specularMaterial()'
-      );
-    });
```
i'm really not sure how to do this. one large part of this PR is the fact that state is no longer held in the shader, and that the shader is selected and updated immediately before drawing. the whole concept of a currently selected shader doesn't exist any more.

kjhollen replied:
I think your logic for selecting the shader in the renderer should be abstracted into a method (or two, if a separate one is needed for immediate mode) and then you can test that it would return the appropriate value after various calls to fill(), stroke(), normalMaterial(), etc.

Spongman commented on test/unit/webgl/stroke.js:
```diff
@@ -20,14 +20,6 @@ suite('stroke WebGL', function() {
-      assert(
-        myp5._renderer.curFillShader === myp5._renderer._getColorShader(),
-        'default fill shader was not initialized with GL canvas'
-      );
```
again, the whole concept of a currently selected shader doesn't exist any longer.

kjhollen replied:
I get that the instance of a particular 'current' shader isn't being stored as a member variable in the same way that it was previously—but I think testing that the right shader is selected given current rendering needs is important to do. For example, a method in the GL Renderer could be:

```
p5.RendererGL.prototype._selectFillShader = function () {
      var fill = this.curFillShader;
      if (this._enableNormal) {
        fill = this._getNormalShader();
      } else if (this._enableLighting) {
        if (!fill || !fill.isLightShader()) {
          fill = this._getImmediateLightShader();
        }
      } else {
        if (!fill /*|| !fill.isColorShader()*/) {
          fill = this._getImmediateFlatShader();
        }
      }
      return fill;
}
```
which is easily testable!

Spongman replied:
okay... but is the test case just a copy of this code?

Spongman commented:
is there anything left to do for this?

kjhollen commented:
Yes. I need to re-review changes after merging with the mainline and also review additional features added since the last review. I've been swamped with organizing a new class and wrapping up other projects; I hope I can get it to it over the weekend or later next week.

Spongman commented:
oh no worries. i was just wondering if you were waiting on something else from me.

kjhollen commented:
hi @Spongman, I've just wrapped up a project and am able to devote more time to reviewing and testing this next week.

I was just reading over your notes about what had been changed/updated since the last review and what hadn't been changed. I still feel that the light shader changes should be in another PR because I think there's more to discuss there. It would be great not to further stall the the rest of this patch while that discussion happens.

My expectation is that by making the immediate and retained mode rendering more aligned, there wouldn't need to be two shaders. There's more to discuss about the specifics for why and how, so I'd prefer to wait and merge (or further refactor) that part until we have a better sense of goals for the lighting architecture. We can open an issue to track that and have a wider discussion, including our two google summer of code participants who will be working on web GL this summer.

Are you willing to update the patch to remove the split of the lighting shader into multiple / additional files? I can help on this next week, too, if needed.

Spongman commented:
i think that even though much of the rendering path of the 'retained' and 'immediate' modes can be combined like this, there is still a need for separate shaders since the set of vertex attributes are quite different for each mode.

for example, the immediate-mode vertices have more attributes that are configurable on a per-vertex level (such as material properties). for retained-mode these properties are per-geometry uniforms. This is also relevant for future per-vertex immediate `tint`, `shininess` & `emissive` support from Processing that i had in the original PR. If you diff `immediateLight.vert` and `light.vert` this should be clear.

I guess it could be possible to add some kind of conditional shader compilation framework, but in the end i think that this would be overkill for what is just two variants - immediate & retained.

Spongman commented:
another thing i was thinking is that this PR makes possible adding a simple geometry builder api.

for example, instead of doing an immediate-mode render, like this:
```javascript
beginShape(TRIANGLE_STRIP);
vertex(-1, -1, 0);
vertex(1, -1, 0);
vertex(-1, 1, 0);
vertex(1, 1, 0);
endShape(CLOSE);
```

you could do something like this:
```javascript
let g = createShape(TRIANGLE_STRIP);
g.vertex(-1, -1, 0);
g.vertex(1, -1, 0);
g.vertex(-1, 1, 0);
g.vertex(1, 1, 0);
g.endShape(CLOSE);
```

and then render that geometry repeatedly:
```javascript
for (...) {
  push();
  translate(...);
  rotate(...);
  render(g);
  pop();
}
```

this would solve the issue that currently the retained-mode stuff is rather opaque and tricky to use outside the simple primitives, and the immediate-mode stuff doesn't perform will when rendering multiple instances of the same geometry.

anyway, food for thought.

kjhollen commented:
Right, I understand why immediate and retain mode lighting are different right now, but what I am suggesting is that may not be true in the long run. we have an open (but old) thread discussing whether the per-vertex fill should be removed as a feature: https://github.com/processing/p5.js/issues/2022, which I'd like to resolve before making a deep refactor to the lighting shader architecture as proposed here.

however, there are many other bug fixes and changes in this PR, which is still why I'd like to merge a portion of it, but updated to minimize the changes to the lighting shaders. if you're willing to do that, I'll review this again when you've updated. otherwise, let's close this for now and I can take on adjusting the patch to minimize the lighting changes. let me know what you'd like to do. thanks!

Spongman commented:
Well, I think that's a shame. There's really no technical reason for disabling per-vertex attributes in WebGL mode, especially since this or makes it so easy to implement. I have a branch with full support for all the processing.org per-vertex attribute with no appreciable performance hit.

I can understand that this was a concern I the past, but I don't thing those concerns are relevant any more and it seems a shame to use them to discard functionality that could easily be added in the near future.

kjhollen commented:
More discussion is needed, as I've said repeatedly—taking it out of this particular patch merely allows the discussion to continue, while accepting some much needed improvements in the meantime!

kjhollen commented:
And either way, please let me know what you'd like to do to push this forward, or if you'd prefer to have me work on it—I really do want to get the bug fixes in, but they're tied up with features that have open issues/discussions expressing a desire to move in the opposite direction. It's not impossible for that to be changed, but we shouldn't hinge the bug fixes here on a related but separate issue that will take longer to discuss and resolve.

Spongman commented:
ok, doing this would mean that #2750, #2274 would no longer be closed by this PR.

kjhollen commented:
ok, that's fine, this still fixes plenty of other issues! :)

Spongman commented:
ok, i took a look at merging the shaders, but it's a mess - it's beyond my patience threshold turning perfect good functional code into a hacked-together mess. maybe someone else can take a stab at it.

kjhollen commented:
Ok, I'll look at it next week when I have time and will close this PR for now. Thanks.

Spongman commented:
are there any plans to merge the rest of this? it seems a shame to let all these fixes and features just die on the vine here...

kjhollen commented:
#3076 is the first step and still waiting on a thorough review. unfortunately, it wasn't realistic to do the work over the last couple months while we had a lot of flux with the GL modules for Google Summer of Code. now that things are a little more settled, it seems like a good time to pick things back up.

for additional features or bugfixes, they still need to be separated into smaller chunks with issues filed for each. anyone is welcome to take on appropriately sized portions of this work.

can you also please file issues as feature requests for any lighting or text changes you're currently thinking of? thank you!
